src/WebJobs.Extensions.DurableTask/DurableTaskExtension.cs

```diff
@@ -464,85 +465,92 @@ protected internal virtual DurableOrchestrationClient GetClient(OrchestrationCli
+                this.TraceHelper.ExtensionInformationalEvent(
+                    this.Options.HubName,
+                    instanceId: string.Empty,
+                    functionName: orchestratorFunction.Name,
+                    message: $"Deregistered orchestrator function named {orchestratorFunction}.",
+                    writeToUserLogs: false);
+            }
         }
 
-        internal OrchestratorInfo GetOrchestratorInfo(FunctionName orchestratorFunction)
+        internal RegisteredFunctionInfo GetOrchestratorInfo(FunctionName orchestratorFunction)
         {
-            OrchestratorInfo info;
-            this.registeredOrchestrators.TryGetValue(orchestratorFunction, out info);
+            RegisteredFunctionInfo info;
+            this.knownOrchestrators.TryGetValue(orchestratorFunction, out info);
 
             return info;
         }
 
         internal void RegisterActivity(FunctionName activityFunction, ITriggeredFunctionExecutor executor)
         {
-            // Allow adding with a null key and subsequently updating with a non-null key.
-            if (!this.registeredActivities.TryUpdate(activityFunction, executor, null))
+            if (this.knownActivities.TryGetValue(activityFunction, out RegisteredFunctionInfo existing))
+            {
+                existing.Executor = executor;
+            }
+            else
             {
                 this.TraceHelper.ExtensionInformationalEvent(
                     this.Options.HubName,
                     instanceId: string.Empty,
                     functionName: activityFunction.Name,
-                    message: $"Registering orchestrator function named {activityFunction}.",
+                    message: $"Registering activity function named {activityFunction}.",
                     writeToUserLogs: false);
 
-                if (!this.registeredActivities.TryAdd(activityFunction, executor))
-                {
-                    throw new ArgumentException($"The activity function named '{activityFunction}' is already registered.");
-                }
+                var info = new RegisteredFunctionInfo(executor, isOutOfProc: false);
+                this.knownActivities[activityFunction] = info;
             }
         }
 
         internal void DeregisterActivity(FunctionName activityFunction)
         {
-            this.TraceHelper.ExtensionInformationalEvent(
-                this.Options.HubName,
-                instanceId: string.Empty,
-                functionName: activityFunction.Name,
-                message: $"Deregistering orchestrator function named {activityFunction}.",
-                writeToUserLogs: false);
+            RegisteredFunctionInfo info;
+            if (this.knownActivities.TryGetValue(activityFunction, out info))
+            {
+                info.IsDeregistered = true;
 
-            this.registeredActivities.TryRemove(activityFunction, out _);
+                this.TraceHelper.ExtensionInformationalEvent(

```

Comment:
NIT: To keep parity with the old behavior, we should technically only log this if the known activity is transitioning from `IsDeregistered` from `false` to `true`.

Reply:
This also applies to the `DerigisterOrchestrator()` logic as well.

Reply:
Fair enough. I'll add a check to see whether it's already been deregistered.